Improve the `OpenAPI Support` Learn page (#4162)
ballerina-platform/ballerina-dev-website: learn/guides/ballerina-openapi-support.md

  - /learn/cli-documentation/openapi/#openapi-validator-compiler-plugin
  - /learn/cli-documentation/openapi/
  - /learn/cli-documentation/openapi
--- 

> Ballerina OpenAPI tooling will make it easy for you to start the development of a service documented in an OpenAPI contract in Ballerina by generating a Ballerina service and client skeletons. It enables you to take the code-first API design approach by generating an OpenAPI contract for the given service implementation.

Author: Since we are referring to Ballerina OpenAPI tooling in general here, we don't need to have "the".

The Ballerina OpenAPI tooling support provides the following capabilities.

> **Prerequisite:** Download and install the Ballerina Swan Lake from [downloads](https://ballerina.io/downloads/).

 1. Generating Ballerina service/client stubs from a given OpenAPI contract file using the CLI tool.
 2. Exporting the OpenAPI definition from a given Ballerina service implementation using the CLI tool.
 3. Validating the service implementation compliance with a provided OpenAPI contract using the OpenAPI annotation.
- The OpenAPI compiler plugin will allow you to validate a service implementation against an OpenAPI contract during the compile time. This plugin ensures that the implementation of a service does not deviate from its OpenAPI contract. 

> ## Generate a Ballerina service 

Reviewer: +1 for option 2, it will give more sense to the user about the content

> ## Generate a Ballerina service 

Author: Option 2 is too lengthy for the recommended size of a title. Option 1 is just the plural form of what we have now. Therefore, it is fine to have the singular form.

> By following an **API design-first approach**, you can generate only the Ballerina service from an OpenAPI definition using the CLI command of the OpenAPI tool as follows.

Reviewer: +1

> By following an **API design-first approach**, you can generate only the Ballerina service from an OpenAPI definition using the CLI command of the OpenAPI tool as follows.

Reviewer: ```suggestion
If you are an API developer who prefers the **design-first approach**, you can use an existing or your own OpenAPI definition to generate Ballerina services using the OpenAPI CLI command, as follows.
```

> By following an **API design-first approach**, you can generate only the Ballerina service from an OpenAPI definition using the CLI command of the OpenAPI tool as follows.

Author: @lnash94 - Please approve.

> By following an **API design-first approach**, you can generate only the Ballerina service from an OpenAPI definition using the CLI command of the OpenAPI tool as follows.

Reviewer: + 1

> By following an **API design-first approach**, you can generate only the Ballerina service from an OpenAPI definition using the CLI command of the OpenAPI tool as follows.

Author: Changed.

-- hello_service.bal
-- types.bal
```
### Generate from tags
To generate the Ballerina service stub with a subset of tags defined in an OpenAPI contract, use the `--tags` option and specify the tags you need as specified in the OpenAPI definition.

```bash
$ bal openapi -i <openapi-contract> [--tags <"tag1","tag2">]
```

For example, 

```bash
$ bal openapi -i hello.yaml --tags "pets", "list"
```

Once you execute the command, only the operations related to the given tags will be included in the generated service file.

>**Info:** For more command options, see [Command reference](#command-reference).

> ## Export an OpenAPI contract

Reviewer: +1 for option 2

> ## Export an OpenAPI contract

Reviewer: @praneesha Could you please verify whether this title is lengthy? it would be great if we can have this 🙂

> ## Export an OpenAPI contract

Author: +1 for option 2. @lnash94 @NipunaRanasinghe - Should it be "Generate" or "Export"?

> ## Export an OpenAPI contract

Reviewer 2: +1 for `export`

> By following an **API code-first approach**, you can convert your Ballerina service APIs into human-readable or machine-readable documents such as OpenAPI documents by using the Ballerina to OpenAPI CLI Tool as follows.

Reviewer: +1 for suggested content

> By following an **API code-first approach**, you can convert your Ballerina service APIs into human-readable or machine-readable documents such as OpenAPI documents by using the Ballerina to OpenAPI CLI Tool as follows.

Author: @lnash94 - Please confirm.

> By following an **API code-first approach**, you can convert your Ballerina service APIs into human-readable or machine-readable documents such as OpenAPI documents by using the Ballerina to OpenAPI CLI Tool as follows.

Reviewer: +1

> By following an **API code-first approach**, you can convert your Ballerina service APIs into human-readable or machine-readable documents such as OpenAPI documents by using the Ballerina to OpenAPI CLI Tool as follows.

Author: Changed.

This will generate the OpenAPI contracts for the Ballerina service in the `hello_service.bal` Ballerina file
of which the `absolute-resource-path` is `/hello`. 
### Export with a given title and version

You can use the `@openapi:ServiceInfo` annotation for specifying the title and version information of the OpenAPI contract as follows.

>**Info:** These `contract`, `title`, and `'version` are all optional attributes.

```ballerina
@openapi:ServiceInfo {
    contract: "/path/to/openapi.json|yaml",
    title: "Store Management",

>     'version: "0.1.0"

Reviewer: Yeah, in line 126 under the info section, it has been mentioned as optional

>     'version: "0.1.0"

Author: @lnash94 - Shall we add comments in the code also for more clarity?

>     'version: "0.1.0"

Author: @NipunaRanasinghe - Please confirm.

## OpenAPI validator compiler plugin

The OpenAPI Validator Compiler plugin validates a service against a given OpenAPI contract. The Compiler Plugin gets activated if a service has the `@openapi:ServiceInfo` annotation. This plugin compares the service and the OpenAPI contract and validates both against a pre-defined set of validation rules. If any of the rules fail, the plugin provides compilation errors.

> 

Author: Should be a mistake. @lnash94 - Please confirm.

> 

Reviewer: yeah, Shall we add this ?

> 

Author: It was removed as it is redundant considering the content. @NipunaRanasinghe - Please check.

> 

Reviewer 2: +1 to remove

The `@openapi:ServiceInfo` annotation is used to bind the service with an OpenAPI contract. You need to add this annotation to the service file with the required values for enabling the validations.

>**Note:** Providing a `contract` path attribute is mandatory for the OpenAPI validator. The other attributes are optional.

The following is an example of the annotation usage in the Ballerina file.

```ballerina
import ballerina/openapi;

@openapi:ServiceInfo {
    contract: "/path/to/openapi.json|yaml",
    tags: ["store"],
    operations: ["op1", "op2"],

>     failOnErrors: true/false → default: true,

Reviewer: ```suggestion
    failOnErrors: true // (default value => true),
```

    excludeTags: ["pets", "user"],
    excludeOperations: ["op1", "op2"]
}
service /greet on new http:Listener(9090) {
  ...
}
```

For annotation attributes details, see [Annotation reference](#annotation-reference).

> ## Generate a Ballerina client

Reviewer: +1 for the suggestion, with make consistence with service title  **Generate Ballerina clients from OpenAPI contract**

> ## Generate a Ballerina client

Author: Changed to `Generate Ballerina clients from OpenAPI definitions`.

-- types.bal
-- utils.bal
```
### Generate with boiler-plate tests

Use the `--with-tests` flag in the client mode to generate a Ballerina client with boilerplate test cases for all the remote functions available in it.
```bash
$ bal openapi -i <openapi-contract> [--mode client] [--with-tests]
```

For example,

```bash
$ bal openapi -i hello.yaml --mode client --with-tests
```
In addition to the above-mentioned generated file, this will generate a `test.bal` file in the default client generation.

### Generate with nillable types

>**Info:** This is an optional flag in the OpenAPI to Ballerina command.

If your OpenAPI specification includes JSON schema properties that are not marked as **nullable:true**, they may be returned as null in some responses which will result in a JSON schema to Ballerina record data binding error. If you suspect this can happen for any property, it is safe to generate all data types in the generated record with Ballerina nil support by turning this flag on.

```bash
$ bal openapi -i <openapi-contract> [-n |--nullable]
```

> For more command options, see [Command reference](#command-reference).

Reviewer 2: if we have a separate reference for `Ballerina to OpenAPI command`, we better not rename this with a generic name like this. @lnash94 WDYT?

Reviewer: Yeah , we have separate command `Ballerina to OpenAPI command`, @praneesha is it possible to keep the existing title since this section contains `OpenAPI to Ballerina command reference` 🙂

Author: I think we should have 2 topics as `Ballerina to OpenAPI command` and `OpenAPI to Ballerina command` and move both reference topics to the References Learn section.

  This is an optional flag in the OpenAPI to Ballerina command. If your OpenAPI specification includes JSON schema properties that are not marked as **nullable:true**, they may return as null in some responses. It will result in a JSON schema to Ballerina record data binding error. If you suspect this can happen for any property, it is safe to generate all data types in the generated record with Ballerina nil support by turning on this flag.

  ```bash
  $ bal openapi -i <openapi-contract> [-n |--nullable]
  ```
  <br/>

**`--with-tests`**<br/>
  This is optional. It works with the client generation command and generates a boiler-plate test for all the remote functions of the generated client.

## Annotation reference

The `@openapi:ServiceInfo` annotation supports several usages in the Ballerina OpenAPI tool. It contains attributes such as `contract` , `tags`, `operations`, `failOnErrors`, `excludeTags`, `excludeOperations`, `title`, `version`, and `embed` for each particular purpose. These attributes are optional to be used in the annotation.

```ballerina
@openapi:ServiceInfo {
    contract: "/path/to/openapi.json|yaml",
    tags: ["store"],
    operations: ["op1", "op2"],

>     failOnErrors: true/false → default: true,

Reviewer 2: ```suggestion
    failOnErrors: true // (default value => true),
```

    excludeTags: ["pets", "user"],
    excludeOperations: ["op1", "op2"],
    title: "store",
    'version: "0.1.0",

>     embed: false/true → default:true

Reviewer 2: ```suggestion
    embed: true // (default value => true)
```
